File: svws-core/src/main/java/de/svws_nrw/core/utils/klausurplan/KlausurblockungSchienenAlgorithmusGreedy3.java
package de.svws_nrw.core.utils.klausurplan;

import java.util.Random;

import jakarta.validation.constraints.NotNull;

/**
 * Die Klausuren werden rekursiv mit Backtracking auf die Schienen verteilt.
 * Pro Rekursionsschritt wird die freie Klausur gewählt, die die meisten Nachbarsfarben hat.
 * Anschließend wird die Klausur in aufsteigender Reihenfolge auf die Schienen verteilt.
 *
 * @author Benjamin A. Bartsch
 */
public final class KlausurblockungSchienenAlgorithmusGreedy3 extends KlausurblockungSchienenAlgorithmusAbstract {

	/** Die kleinste Schienenanzahl, die bisher gefunden wurde. */
	private int _minSchienen;

	/** Bis zu welchem Zeitpunkt die Rekursion laufen darf. */
	private long _zeitEnde;

	/** TRUE, falls mindestens eine Lösung gefunden wurde. */
	private boolean _saved;

	/** Konstruktor.
	 *
	 * @param pRandom   Ein {@link Random}-Objekt zur Steuerung des Zufalls über einen Anfangs-Seed.
	 * @param pDynDaten Die aktuellen Blockungsdaten. */
	public KlausurblockungSchienenAlgorithmusGreedy3(final @NotNull Random pRandom, final @NotNull KlausurblockungSchienenDynDaten pDynDaten) {
		super(pRandom, pDynDaten);
	}

	@Override
	public @NotNull String toString() {
		return "Backtracking";
	}

	@Override
	public void berechne(final long pZeitEnde) {
		_minSchienen = _dynDaten.gibAnzahlKlausuren();
		_zeitEnde = pZeitEnde;
		_dynDaten.aktionKlausurenAusSchienenEntfernen();
		_saved = false;

		berechneRekursiv();

		_dynDaten.aktionZustand1Laden();

		if (_dynDaten.gibIstBesserAlsZustand2() == true)
			_dynDaten.aktionZustand2Speichern();
	}

	private void berechneRekursiv() {

		// Kann das Ergebnis überhaupt noch besser werden?
		if (_dynDaten.gibAnzahlSchienen() > _minSchienen)
			return;

		// Zeit abgelaufen (aber mindestens eine Speicherung)?
		if ((_saved) && (System.currentTimeMillis() > _zeitEnde))
			return;

		// Erster Aufruf?
		final int klausurNr = _dynDaten.gibAnzahlSchienen() == 0 ? //
				_dynDaten.gibKlausurDieFreiIstMitDenMeistenFreienNachbarn() : // 1.
				_dynDaten.gibKlausurDieFreiIstMitDenMeistenNachbarsfarben(); // 2+.

		// Alle Klausuren gesetzt?
		if (klausurNr < 0) {
			if (!_saved || _dynDaten.gibIstBesserAlsZustand1()) {
				_minSchienen = _dynDaten.gibAnzahlSchienen();
				_dynDaten.aktionZustand1Speichern();
				_saved = true;
			}
			return;
		}

		// Rekursion
		for (int schiene = 0; schiene < _minSchienen; schiene++) {
			final int schienenAnzahl = _dynDaten.gibAnzahlSchienen();
			final int differenz = schiene < schienenAnzahl ? 0 : (schiene - schienenAnzahl + 1);

			_dynDaten.aktionSchienenAnzahlVeraendern(+differenz);
			if (_dynDaten.aktionSetzeKlausurInSchiene(klausurNr, schiene) == true) {
				berechneRekursiv();
				_dynDaten.aktionEntferneKlausurAusSchiene(klausurNr);
			}
			_dynDaten.aktionSchienenAnzahlVeraendern(-differenz);
		}

	}

}

File: svws-webclient/core/src/core/utils/klausurplanung/KlausurblockungSchienenAlgorithmusGreedy3.ts
import { KlausurblockungSchienenAlgorithmusAbstract } from '../../../core/utils/klausurplanung/KlausurblockungSchienenAlgorithmusAbstract';
import { KlausurblockungSchienenDynDaten } from '../../../core/utils/klausurplanung/KlausurblockungSchienenDynDaten';
import { Random } from '../../../java/util/Random';
import { System } from '../../../java/lang/System';

export class KlausurblockungSchienenAlgorithmusGreedy3 extends KlausurblockungSchienenAlgorithmusAbstract {

	/**
	 * Die kleinste Schienenanzahl, die bisher gefunden wurde.
	 */
	private _minSchienen : number = 0;

	/**
	 * Bis zu welchem Zeitpunkt die Rekursion laufen darf.
	 */
	private _zeitEnde : number = 0;

	/**
	 * TRUE, falls mindestens eine Lösung gefunden wurde.
	 */
	private _saved : boolean = false;


	/**
	 *Konstruktor.
	 *
	 * @param pRandom   Ein {@link Random}-Objekt zur Steuerung des Zufalls über einen Anfangs-Seed.
	 * @param pDynDaten Die aktuellen Blockungsdaten.
	 */
	public constructor(pRandom : Random, pDynDaten : KlausurblockungSchienenDynDaten) {
		super(pRandom, pDynDaten);
	}

	public toString() : string {
		return "Backtracking";
	}

	public berechne(pZeitEnde : number) : void {
		this._minSchienen = this._dynDaten.gibAnzahlKlausuren();
		this._zeitEnde = pZeitEnde;
		this._dynDaten.aktionKlausurenAusSchienenEntfernen();
		this._saved = false;
		this.berechneRekursiv();
		this._dynDaten.aktionZustand1Laden();
		if (this._dynDaten.gibIstBesserAlsZustand2())
			this._dynDaten.aktionZustand2Speichern();
	}

	private berechneRekursiv() : void {
		if (this._dynDaten.gibAnzahlSchienen() > this._minSchienen)
			return;
		if ((this._saved) && (System.currentTimeMillis() > this._zeitEnde))
			return;
		const klausurNr : number = this._dynDaten.gibAnzahlSchienen() === 0 ? this._dynDaten.gibKlausurDieFreiIstMitDenMeistenFreienNachbarn() : this._dynDaten.gibKlausurDieFreiIstMitDenMeistenNachbarsfarben();
		if (klausurNr < 0) {
			if (!this._saved || this._dynDaten.gibIstBesserAlsZustand1()) {
				this._minSchienen = this._dynDaten.gibAnzahlSchienen();
				this._dynDaten.aktionZustand1Speichern();
				this._saved = true;
			}
			return;
		}
		for (let schiene : number = 0; schiene < this._minSchienen; schiene++) {
			const schienenAnzahl : number = this._dynDaten.gibAnzahlSchienen();
			const differenz : number = schiene < schienenAnzahl ? 0 : (schiene - schienenAnzahl + 1);
			this._dynDaten.aktionSchienenAnzahlVeraendern(+differenz);
			if (this._dynDaten.aktionSetzeKlausurInSchiene(klausurNr, schiene)) {
				this.berechneRekursiv();
				this._dynDaten.aktionEntferneKlausurAusSchiene(klausurNr);
			}
			this._dynDaten.aktionSchienenAnzahlVeraendern(-differenz);
		}
	}

	isTranspiledInstanceOf(name : string): boolean {
		return ['de.svws_nrw.core.utils.klausurplanung.KlausurblockungSchienenAlgorithmusGreedy3', 'de.svws_nrw.core.utils.klausurplanung.KlausurblockungSchienenAlgorithmusAbstract'].includes(name);
	}

}

export function cast_de_svws_nrw_core_utils_klausurplanung_KlausurblockungSchienenAlgorithmusGreedy3(obj : unknown) : KlausurblockungSchienenAlgorithmusGreedy3 {
	return obj as KlausurblockungSchienenAlgorithmusGreedy3;
}
